_posts/2021-08-12-deep-dive-into-iceberg-internals.md

```diff
@@ -0,0 +1,770 @@
+is akin to what you would see in a Hive table data directory. What is really of
+interest here is the metadata directory. There are specifically three patterns
+of files you’ll find here.
+
+`/<bucket>/<database>/<table>/metadata/<file-id>.avro/<bucket>/<database>/<table>/metadata/snap-<snapshot-id>-<version>-<file-id>.avro`
+
+`/<bucket>/<database>/<table>/metadata/<version>-<commit-UUID>.metadata.json`
+
+Iceberg has a persistent tree structure that manages various snapshots of the
+data that are created for every mutation of the data. This enables not only a
+concurrency model that supports serializable isolation, but also cool features
+like time travel across a linear progression of snapshots.
+
+![](/assets/blog/trino-on-ice/iceberg-metastore-files.png)
+
+This tree structure contains two types of avro files, manifest lists and
+manifest files. Manifest list files contain pointers to various manifest files
+and the manifest files themselves point to various data files. This blog starts
+out by covering these manifest files, and later covers the table metadata files
+that are suffixed by `.metadata.json`.
+
+[The last blog covered]({% post_url 2021-07-30-iceberg-concurrency-snapshots-spec %})
+the command in Trino that shows the
+snapshot information that is stored in the metastore. Here is that command and
+its output again for your review.
+
+```
+SELECT manifest_list 
+FROM iceberg.logging."events$snapshots";
+```
+
+Result:
+
+| snapshots |
+| --- |
+| s3a://iceberg/logging.db/events/metadata/snap-6967685587675910019-1-bcbe9133-c51c-42a9-9c73-f5b745702cb0.avro |
+| s3a://iceberg/logging.db/events/metadata/snap-2720489016575682283-1-92382234-a4a6-4a1b-bc9b-24839472c2f6.avro |
+| s3a://iceberg/logging.db/events/metadata/snap-4564366177504223943-1-23cc980c-9570-42ed-85cf-8658fda2727d.avro |
+
+You’ll notice that the manifest list returns the Avro files prefixed with
+`snap-` are returned. These files are directly correlated with the snapshot
+record stored in the metastore. According to the diagram above, snapshots are
+records in the metastore that contain the url of the manifest list in the Avro
+file. Avro files are binary files and not something you can just open up in a
+text editor to read. Using the 
+[avro-tools.jar tool](https://downloads.apache.org/avro/avro-1.10.2/java/avro-tools-1.10.2.jar)
+distributed by the 
+[Apache Avro project](https://avro.apache.org/docs/current/index.html),
+you can actually inspect the contents of this file to get a better understanding
+of how it is used by Iceberg.
+
+The first snapshot is generated on the creation of the events table. Upon
+inspecting this file, you will notice that the file is empty. The output is an
+empty line that the jq json command line utility removes on pretty printing the
+json that is returned, which is just a newline. This snapshot represents an
+empty state of the table upon creation. In order to investigate these snapshots,
```

Comment:
Cut the clutter

To investigate the snapshots you need to download the files to your local filesystem. Let's move them to the home directory: